src/peripheral/mod.rs

```diff
@@ -484,6 +545,230 @@ impl Scb {
+        // All of CBP is write-only so no data races are possible
+        let cbp = unsafe { &mut *CBP.get() };
+
+        ::asm::dsb();
+        ::asm::isb();
+
+        // Invalidate I-Cache
+        cbp.iciallu();
+
+        ::asm::dsb();
+        ::asm::isb();
+    }
+
+    /// Enables D-cache
+    #[cfg(armv7m)]
+    #[inline]
+    pub fn enable_dcache(&self, cpuid: &Cpuid) {
+        // Invalidate anything currently in the DCache
+        self.invalidate_dcache(cpuid);
+
+        // Now turn on the DCache
+        unsafe { self.ccr.modify(|r| r | SCB_CCR_DC_MASK) };
+
+        ::asm::dsb();
+        ::asm::isb();
+    }
+
+    /// Disables D-cache
+    #[cfg(armv7m)]
+    #[inline]
+    pub fn disable_dcache(&self, cpuid: &Cpuid) {
+        // Turn off the DCache
+        unsafe { self.ccr.modify(|r| r & !SCB_CCR_DC_MASK) };
+
+        // Clean and invalidate whatever was left in it
+        self.clean_invalidate_dcache(cpuid);
+    }
+
+    /// Invalidates D-cache
+    #[cfg(armv7m)]
+    #[inline]
+    pub fn invalidate_dcache(&self, cpuid: &Cpuid) {
+        // All of CBP is write-only so no data races are possible
+        let cbp = unsafe { &mut *CBP.get() };
+
+        // Read number of sets and ways
+        let sets = cpuid.cache_num_sets(0, 0);
+        let ways = cpuid.cache_num_ways(0, 0);
+
+        // Invalidate entire D-Cache
+        for set in 0..sets {
+            for way in 0..ways {
+                cbp.dcisw(set, way);
+            }
+        }
+
+        ::asm::dsb();
+        ::asm::isb();
+    }
+
+    /// Cleans D-cache
+    #[cfg(armv7m)]
+    #[inline]
+    pub fn clean_dcache(&self, cpuid: &Cpuid) {
+        // All of CBP is write-only so no data races are possible
+        let cbp = unsafe { &mut *CBP.get() };
+
+        // Read number of sets and ways
+        let sets = cpuid.cache_num_sets(0, 0);
+        let ways = cpuid.cache_num_ways(0, 0);
+
+        for set in 0..sets {
+            for way in 0..ways {
+                cbp.dccsw(set, way);
+            }
+        }
+
+        ::asm::dsb();
+        ::asm::isb();
+    }
+
+    /// Cleans and invalidates D-cache
+    #[cfg(armv7m)]
+    #[inline]
+    pub fn clean_invalidate_dcache(&self, cpuid: &Cpuid) {
+        // All of CBP is write-only so no data races are possible
+        let cbp = unsafe { &mut *CBP.get() };
+
+        // Read number of sets and ways
+        let sets = cpuid.cache_num_sets(0, 0);
+        let ways = cpuid.cache_num_ways(0, 0);
+
+        for set in 0..sets {
+            for way in 0..ways {
+                cbp.dccisw(set, way);
+            }
+        }
+
+        ::asm::dsb();
+        ::asm::isb();
+    }
+
+    /// Invalidates D-cache by address
+    ///
+    /// `addr`: the address to invalidate, aligned to 32-byte boundary
+    /// `size`: size of the memory block, in number of bytes, a multiple of 32
+    #[cfg(armv7m)]
+    #[inline]
+    pub fn invalidate_dcache_by_address(&self, addr: u32, size: u32) {
```

Comment:
The requirements stated in the documentation should be asserted in the body of this method, otherwise (I suppose) Bad Stuff could happen if the user calls this with non aligned address and sizes not multiple of 32.

Alternatively you could say that `size` has units of "lines" and a `size` of 1 equals to 32 bytes.

Same comment for the methods below.